azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -0,0 +1,411 @@
+        Fetches the goal state using the given wire client.
+
+        By default it fetches only the goal state itself; to fetch the entire goal state (that includes all the
+        nested components, such as the extension config) use the 'full_goal_state' parameter.
+
+        If 'base_incarnation' is given, it fetches the full goal state if the new incarnation is different than
+        the given value, otherwise it fetches only the goal state itself.
+
+        For better code readability, use the static fetch_* methods below instead of instantiating GoalState
+        directly.
+
+        """
+        uri = GOAL_STATE_URI.format(wire_client.get_endpoint())
+        self.xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+        xml_doc = parse_doc(self.xml_text)
+
+        self.incarnation = findtext(xml_doc, "Incarnation")
+        self.expected_state = findtext(xml_doc, "ExpectedState")
+        role_instance = find(xml_doc, "RoleInstance")
+        self.role_instance_id = findtext(role_instance, "InstanceId")
+        role_config = find(role_instance, "Configuration")
+        self.role_config_name = findtext(role_config, "ConfigName")
+        container = find(xml_doc, "Container")
+        self.container_id = findtext(container, "ContainerId")
+        lbprobe_ports = find(xml_doc, "LBProbePorts")
+        self.load_balancer_probe_port = findtext(lbprobe_ports, "Port")
+
+        os.environ[CONTAINER_ID_ENV_VARIABLE] = self.container_id
+
+        if not (full_goal_state or base_incarnation is not None and self.incarnation != base_incarnation):
+            self.hosting_env = None
+            self.shared_conf = None
+            self.certs = None
+            self.ext_conf = None
+            self.remote_access = None
+            return
+
+        uri = findtext(xml_doc, "HostingEnvironmentConfig")
+        xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+        self.hosting_env = HostingEnv(xml_text)
+
+        uri = findtext(xml_doc, "SharedConfig")
+        xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+        self.shared_conf = SharedConfig(xml_text)
+
+        uri = findtext(xml_doc, "Certificates")
+        if uri is None:
+            self.certs = None
+        else:
+            xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
+            self.certs = Certificates(xml_text)
+
+        uri = findtext(xml_doc, "ExtensionsConfig")
+        if uri is None:
+            self.ext_conf = ExtensionsConfig(None)
+        else:
+            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+            self.ext_conf = ExtensionsConfig(xml_text)
+
+        uri = findtext(container, "RemoteAccessInfo")
+        if uri is None:
+            self.remote_access = None
+        else:
+            xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
+            self.remote_access = RemoteAccess(xml_text)
+
```

Comment:
I'm planning on moving the fetch_* methods below to protocol, and remove the current interface that exposes the goal state from the protocol class. I.e. code would do "goal_state = protocol.fetch_goal_state()" and then use "goal_state" directly instead of using "protocol" to retrieve the properties of the goal state.